test/src/specs/tx_pool/dao.rs

```diff
@@ -0,0 +1,403 @@
+use ckb_core::transaction::{CellInput, CellOutput, OutPoint, Transaction, TransactionBuilder};
+use ckb_core::{BlockNumber, Bytes, Capacity};
+use ckb_resource::CODE_HASH_DAO;
+use ckb_test_chain_utils::always_success_cell;
+use numext_fixed_hash::H256;
+
+const SYSTEM_CELL_ALWAYS_SUCCESS_INDEX: u32 = 1;
+const SYSTEM_CELL_DAO_INDEX: u32 = 3;
+const WITHDRAW_WINDOW_LEFT: u64 = 10;
+const WITHDRAW_HEADER_INDEX: u8 = 0;
+
+pub struct DepositDAO;
+
+impl Spec for DepositDAO {
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        // Deposit DAO
+        {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction);
+        }
+
+        // Deposit DAO without specifying `block_hash` within input
+        {
+            // deposit dao transaction without `block_hash`
+            let transaction = {
+                let out_point_without_block_hash = {
+                    let tx_hash = node0.get_tip_block().transactions()[0].hash().to_owned();
+                    OutPoint::new_cell(tx_hash, 0)
+                };
+                TransactionBuilder::from_transaction(deposit_dao_transaction(node0))
+                    .inputs_clear()
+                    .input(CellInput::new(out_point_without_block_hash, 0))
+                    .build()
+            };
+            ensure_committed(node0, &transaction);
+        }
+    }
+}
+
+pub struct WithdrawDAO;
+
+impl Spec for WithdrawDAO {
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        let deposited = {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction)
+        };
+        let transaction = withdraw_dao_transaction(node0, deposited.clone());
+        ensure_committed(node0, &transaction);
+    }
+}
+
+pub struct DepositDAOTwice;
+
+impl Spec for DepositDAOTwice {
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        let mut deposited = {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction)
+        };
+        for _ in 0..5 {
+            let transaction = {
+                let transaction = withdraw_dao_transaction(node0, deposited.clone());
+                let outputs: Vec<_> = transaction
+                    .outputs()
+                    .iter()
+                    .cloned()
+                    .map(|mut cell_output| {
+                        cell_output.type_ = Some(deposit_dao_script());
+                        cell_output
+                    })
+                    .collect();
+                TransactionBuilder::from_transaction(transaction)
+                    .outputs_clear()
+                    .outputs(outputs)
+                    .build()
+            };
+            // TODO compare the reward
```

Comment:
I did saw we introduced `calculate_maximum_withdraw` in this PR, so I'm curious what's preventing adding the actual logic here?

Reply:
Nothing prevents this logic here.
It is because of that I didn't think clearly about which aspects of reward I should test. So I didn't compare the reward in this PR and just leave a TODO.
Now it more clear. I will add several reward tests later.